### PR TITLE
Provisioning feature as provision-block in config file

### PR DESCRIPTION
**Pull request by andersla**
## Change content and motivation
Adds provisioning feature (ansible-playbook, helm or bash) as provision-block in config file.

### CODE REVIEW

File: bin/kn-provision [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
+# Exit immediately if a command exits with a non-zero status
+set -e
+
+echo "Inside kn-provision"
```
Comment (mcapuccini) [inline bin/kn-provision]: Please change with something line "Provisioning bla bla...", IMO we shouldn't assume that the user knows the internal structure of the project

File: bin/kn-provision [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# Convert to base64 to preserve whitespace and citations in bash loop
+# Execute actions (ansible or local-exec) with supplied variables/parameters
+actions=$(echo "$provision" | jq -r --compact-output '.action[]? | @base64')
+for action_base64 in $actions; do
+  action=$(base64 --decode <<<"$action_base64")
+
+  type=$(echo "$action" | jq -r '.type')
+  case "$type" in
+  ansible-playbook)
+    echo "Inside kn-provision ansible-playbook"
```
Comment (mcapuccini) [inline bin/kn-provision]: Same comment as before

File: bin/kn-provision [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    extra_vars=$(echo "$action" | jq -r --compact-output 'select(.extra_vars != null) | .extra_vars[]')
+
+    echo "Applying playbook $playbook"
+    # shellcheck disable=SC2086
+    ansible-playbook $ANSIBLE_OPT -e "$extra_vars" "/KubeNow_root/$playbook"
+    echo "Done with ansible-playbook: $playbook"
+    ;;
+
+  local-exec)
+    echo "Inside kn-provision local-exec"
```
Comment (mcapuccini) [inline bin/kn-provision]: Same comment as before

File: bin/kn-provision [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    command=$(echo "$action" | jq -r '.command')
+
+    echo "Exec command: $command"
+
+    # execute command
+    $command
+    echo "Done with local-exec"
+    ;;
```
Comment (mcapuccini) [inline bin/kn-provision]: Maybe add a case to catch unrecognized provisioner types.

**Comment by andersla** [comment]
I guess it is ready to merge?

**Comment by mcapuccini** [comment]
@andersla @carmat88 there are still two echoes to fix, see my comments above. After we sort out these wear ready to go!

File: bin/kn-provision [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  "")
+    echo "Error: No action-type specified in provisioner-block" >&2
+    echo "Exiting"
+    exit 1
+    ;;
+  *)
+    echo "Error: Unknown action-type in provisioner-block: $type" >&2
+    echo "Exiting"
+    exit 1
+    ;;
```
Comment (carmat88) [inline bin/kn-provision]: @mcapuccini Thanks for checking this out. Just to be sure, haven't those two echoes been addressed here in  `bin/kn-provision` on lines 43-52 ? Or am I completely misunderstanding it :-) ?

Comment (mcapuccini) [reply bin/kn-provision]: No, I meant `echo "Inside kn-provision local-exec"`, and sorry it's only one to fix.

**Comment by andersla** [comment]
:-)

**Pull request merged**